### PR TITLE
[TRTLLM-11403][feat] VisualGen Cache-DiT + unified cache accelerator

### PR DESCRIPTION
**Pull request by o-stoner**
## Release Notes

* **New Features**
  * Added Cache-DiT acceleration as an alternative caching option for visual generation pipelines alongside TeaCache
  * Introduced cache backend selection mechanism to choose between TeaCache and Cache-DiT via configuration and CLI flags
  * Supports configurable Cache-DiT parameters including warmup steps, cached steps, and residual thresholds for optimized performance on Flux and Wan models

## Description

Unified cache accelerator and Cache-DiT implementation for Wan, FLUX, and LTX-2 models.

Perf/quality comparison for each cache backend from example scripts:
| Model | Configs | Pipeline Time (Default) | Pipeline Time (TeaCache) | Pipeline Time (Cache-DiT) | Media/Logs Link |
| --- | --- | --- | --- | --- | --- |
| FLUX.2-dev | "A cat sitting on a windowsill", 1024x1024, 50 steps, trtllm-nvfp4 linear, VANILLA attention | 5.60s | **2.20s** | 3.01s | [link](https://sc.talos.nvidia.com/view/home/scratch.ostoner_coreai_1/TensorRT-LLM/examples/visual_gen/perf_tests_cachedit_pr/FLUX.2-dev) |
| Wan2.1-T2V-1.3B-Diffusers | "A cute cat playing piano", 480x832, 33 frames, 50 steps, trtllm-fp8-blockwise linear, VANILLA attention | 7.94s | **2.84s** | 4.79s | [link](https://sc.talos.nvidia.com/view/home/scratch.ostoner_coreai_1/TensorRT-LLM/examples/visual_gen/perf_tests_cachedit_pr/Wan2.1-T2V-1.3B-Diffusers) |
| Wan2.1-I2V-14B-480P-Diffusers | 480x832, 81 frames, 50 steps, trtllm-nvfp4 linear, TRTLLM attention, parallel VAE disabled | 128.79s | **60.61s** | 66.34s | [link](https://sc.talos.nvidia.com/view/home/scratch.ostoner_coreai_1/TensorRT-LLM/examples/visual_gen/perf_tests_cachedit_pr/Wan2.1-I2V-14B-480P-Diffusers) |
| Wan2.1-I2V-14B-720P-Diffusers | 720x1280, 81 frames, 50 steps, trtllm-nvfp4 linear, TRTLLM attention, parallel VAE disabled | 501.23s | 233.56s | **220.34s** | [link](https://sc.talos.nvidia.com/view/home/scratch.ostoner_coreai_1/TensorRT-LLM/examples/visual_gen/perf_tests_cachedit_pr/Wan2.1-I2V-14B-720P-Diffusers) |
| Wan2.1-T2V-14B-Diffusers | "A cute cat playing piano", 720x1280, 81 frames, 50 steps, trtllm-nvfp4 linear, TRTLLM attention | 484.63s | 243.55s | **203.62s** | [link](https://sc.talos.nvidia.com/view/home/scratch.ostoner_coreai_1/TensorRT-LLM/examples/visual_gen/perf_tests_cachedit_pr/Wan2.1-T2V-14B-Diffusers) |
| Wan2.2-I2V-A14B-Diffusers | 720x1280, 81 frames, 40 steps, trtllm-nvfp4 linear, TRTLLM attention, parallel VAE disabled | 387.83s | N/A | **205.28s** | [link](https://sc.talos.nvidia.com/view/home/scratch.ostoner_coreai_1/TensorRT-LLM/examples/visual_gen/perf_tests_cachedit_pr/Wan2.2-I2V-A14B-Diffusers) |
| Wan2.2-T2V-A14B-Diffusers | "A cute cat playing piano", 720x1280, 81 frames, 40 steps, trtllm-nvfp4 linear, TRTLLM attention, parallel VAE disabled | 389.10s | N/A | **189.63s** | [link](https://sc.talos.nvidia.com/view/home/scratch.ostoner_coreai_1/TensorRT-LLM/examples/visual_gen/perf_tests_cachedit_pr/Wan2.2-T2V-A14B-Diffusers) |
| LTX-2 (model: LTX-2-NVFP4, text encoder: gemma-3-12b-it) | "Wide shot of a shimmering, crystal-blue waterfall cascading down a glowing, magical jungle cliffside. Sunlight filters through the bioluminescent trees, casting a vibrant, colorful shimmer on the misty water. The camera slowly zooms in as vibrant birds fly across the frame.", 720x1280, 121 frames, 40 steps, VANILLA attention | 34.23s | N/A | **20.09s** | [link](https://sc.talos.nvidia.com/view/home/scratch.ostoner_coreai_1/TensorRT-LLM/examples/visual_gen/perf_tests_cachedit_pr/LTX2) |

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by o-stoner** [comment]
@zhenhuaw-me @yibinl-nvidia @chang-l do you mind taking a look when you get a chance? WAN, FLUX, and LTX-2 Cache-DiT are now enabled. Thank you in advance!

File: tensorrt_llm/_torch/visual_gen/pipeline.py [hunk 1]
```diff
@@ -259,65 +262,83 @@ def post_load_weights(self) -> None:
         if self.transformer is not None and hasattr(self.transformer, "post_load_weights"):
             self.transformer.post_load_weights()
 
-    def _setup_teacache(self, model, coefficients: Optional[Dict] = None):
-        """Setup TeaCache optimization for the transformer model.
+    def _apply_teacache_coefficients(self, coefficients: Optional[Dict]) -> None:
```
[the remark this answers is not in the document]
Comment (o-stoner) [reply tensorrt_llm/_torch/visual_gen/pipeline.py]: Yes, assigned the task to myself. Will follow up in a later PR

File: tensorrt_llm/_torch/visual_gen/config.py [hunk 2]
```diff
@@ -365,6 +429,11 @@ class VisualGenArgs(StrictBaseModel):
     attention: AttentionConfig = PydanticField(default_factory=AttentionConfig)
     parallel: ParallelConfig = PydanticField(default_factory=ParallelConfig)
     teacache: TeaCacheConfig = PydanticField(default_factory=TeaCacheConfig)
+    cache_backend: CacheBackendName = PydanticField(
+        "none",
+        description="Cache accelerator: none, teacache, or cache_dit.",
+    )
+    cache_dit: CacheDiTConfig = PydanticField(default_factory=CacheDiTConfig)
```
[the remark this answers is not in the document]
Comment (NVShreyas) [reply tensorrt_llm/_torch/visual_gen/config.py]: thanks, lgtm

File: tensorrt_llm/_torch/visual_gen/models/ltx2/transformer_ltx2.py [hunk 3]
```diff
@@ -673,6 +673,49 @@ def forward(
         )
 
 
+class LTX2CacheDiTPattern0BlockWrapper(nn.Module):
```
Comment (chang-l) [inline tensorrt_llm/_torch/visual_gen/models/ltx2/transformer_ltx2.py:677]: It seems the module-level changes to the LTX2 transformer is significant, unlike WAN/Flux. @yibinl-nvidia @luyiyun1021 can you please help review this part?

File: examples/visual_gen/visual_gen_wan_i2v.py [hunk 4]
```diff
@@ -8,6 +8,7 @@
 import time
 
 from tensorrt_llm import VisualGen, VisualGenArgs, VisualGenParams, logger
+from tensorrt_llm._torch.visual_gen.config import CacheDiTConfig, TeaCacheConfig
```
Comment (chang-l) [inline examples/visual_gen/visual_gen_wan_i2v.py:11]: Do we need to update all server config yaml files (`examples/visual_gen/serve/configs/`) to support configurable cache backend?

File: examples/visual_gen/visual_gen_wan_i2v.py [hunk 5]
```diff
@@ -173,16 +241,51 @@ def _linear_type_to_quant_config(linear_type: str):
...
+    kwargs: dict = {"use_ret_steps": args.use_ret_steps}
+    if args.teacache_thresh is not None:
+        kwargs["teacache_thresh"] = args.teacache_thresh
+    return TeaCacheConfig(**kwargs)
+
+
+def _cache_dit_config_from_args(args) -> CacheDiTConfig:
+    """Subset of CacheDiTConfig from CLI; unset options keep Pydantic defaults."""
+    overrides: dict = {}
+    if args.cache_dit_fn_compute_blocks is not None:
```
Comment (chang-l) [inline examples/visual_gen/visual_gen_wan_i2v.py:255]: Nit (non-blocking): The `_cache_dit_config_from_args()` helper is repeated across all example scripts. Could it be extracted into a shared example utility (similar to the LLM examples, we can have `add_cache_args` and `build_cache_config`)? Feel free to address this in a follow-up example refactor PR.
cc. @zhenhuaw-me

**Review by chang-l — APPROVED** [review]

File: tensorrt_llm/_torch/visual_gen/pipeline.py [hunk 6]
```diff
@@ -259,65 +262,83 @@ def post_load_weights(self) -> None:
...
-        TeaCache caches transformer block outputs when timestep embeddings change slowly,
-        reducing computation during the denoising loop.
+    def _setup_cache_acceleration(
+        self,
+        model: Optional[nn.Module] = None,
+        coefficients: Optional[Dict] = None,
+    ) -> None:
+        """Enable TeaCache or Cache-DiT from model_config.cache_backend."""
 
-        Args:
```
[the remark this answers is not in the document]
Comment (zhenhuaw-me) [reply tensorrt_llm/_torch/visual_gen/pipeline.py:307]: I prefer to keep them.

File: tensorrt_llm/_torch/visual_gen/models/ltx2/transformer_ltx2.py (hunk 3 above)
Comment (luyiyun1021) [inline tensorrt_llm/_torch/visual_gen/models/ltx2/transformer_ltx2.py:677]: One concern about Pattern_0 semantics for the AudioVideo case:

Cache-dit's caching decision is based solely on hidden_states (slot 1 = video latent) residual diff. encoder_hidden_states (slot 2 = audio latent) is not factored into the skip decision — it's only passively cached/restored.

This is fine for the original Pattern_0 use case (image + text conditioning, where text barely changes), but for LTX2 AudioVideo mode, both streams are actively being denoised. If video residual happens to be small at some step while audio is still changing significantly, the block would be skipped and audio would get a stale cached approximation (not sure if this may happen).

The bidirectional cross-attention coupling likely keeps the two streams roughly in sync, so this may be acceptable in practice. Have you checked the audio-quality after caching?

Comment (o-stoner) [reply tensorrt_llm/_torch/visual_gen/models/ltx2/transformer_ltx2.py:677]: Understood, thanks for raising the concern. Take a look at the videos [here](https://sc.talos.nvidia.com/view/home/scratch.ostoner_coreai_1/TensorRT-LLM/examples/visual_gen/perf_tests_cachedit_pr/LTX2) (from Media/Logs link above). Seems to be fairly similar audio quality for default and Cache-DiT outputs. This Pattern_0 semantics follow suit with vLLM-Omni's implementation [here](https://github.com/vllm-project/vllm-omni/blob/main/vllm_omni/diffusion/cache/cache_dit_backend.py#L606); they treated audio_hidden_states as encoder_hidden_states in Pattern_0. Let me know your thoughts.

File: examples/visual_gen/visual_gen_wan_i2v.py (hunk 4 above)
Comment (o-stoner) [reply examples/visual_gen/visual_gen_wan_i2v.py:11]: Yes, we do - we will also need to update the TeaCache lines in https://github.com/NVIDIA/TensorRT-LLM/pull/12992 once this is merged (if that PR is not meant to replace the `examples/visual_gen/serve/configs/` files). I am assuming we want to include the best-performing cache strategies in these example configs as opposed to just keeping TeaCache?

File: tensorrt_llm/_torch/visual_gen/models/ltx2/transformer_ltx2.py (hunk 3 above)
Comment (luyiyun1021) [reply tensorrt_llm/_torch/visual_gen/models/ltx2/transformer_ltx2.py:677]: Thanks. LGTM

File: examples/visual_gen/visual_gen_wan_i2v.py (hunk 4 above)
Comment (o-stoner) [reply examples/visual_gen/visual_gen_wan_i2v.py:11]: Updated the `examples/visual_gen/serve/configs/` files in my most recent commit to include the best-performing cache strategies for each model.

**Comment by o-stoner** [comment]
@NVIDIA/trt-llm-oss-compliance do you mind giving this a review when you get a chance? Thank you!

**Review by tburt-nv — APPROVED** [review]
cache-dit is Apache licensed and introduces no new transitive dependencies, approved.
